src/meta/src/barrier/mod.rs

```diff
@@ -415,41 +451,66 @@ impl CheckpointControl {
                         warn!(err = ?e.as_report(), "failed to complete barrier during clear");
                         true
                     }
-                    Ok(Ok(_)) => false,
+                    Ok(Ok(hummock_version_stats)) => {
+                        self.checkpoint_control
+                            .ack_completed(BarrierCompleteOutput {
+                                command_prev_epoch,
+                                creating_job_epochs,
+                                hummock_version_stats,
+                            });
+                        false
+                    }
                 }
             }
             CompletingTask::Err(_) => true,
         };
         if !is_err {
             // continue to finish the pending collected barrier.
-            while let Some(task) = self.next_complete_barrier_task(None) {
-                if let Err(e) = self.context.clone().complete_barrier(task).await {
-                    error!(
-                        err = ?e.as_report(),
-                        "failed to complete barrier during recovery"
-                    );
-                    break;
-                } else {
-                    info!("succeed to complete barrier during recovery")
+            while let Some(task) = self.checkpoint_control.next_complete_barrier_task(None) {
+                let (command_prev_epoch, creating_job_epochs) = (
+                    task.command_context
+                        .as_ref()
+                        .map(|(command, _)| command.barrier_info.prev_epoch.value().0),
+                    task.creating_job_epochs.clone(),
```

Comment:
nit: I think we should provide a getter for this. `command.barrier_info.prev_epoch.value().0` is not very readable.